daemon/containerd/image_builder.go

```diff
@@ -307,22 +372,33 @@ func (rw *rwlayer) Commit() (builder.ROLayer, error) {
 		return nil, err
 	}
 
+	if err := lm.AddResource(ctx, lease, leases.Resource{
+		ID:   desc.Digest.String(),
+		Type: "content",
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add diff content to lease")
+	}
+
 	return &rolayer{
 		key:                key,
 		c:                  rw.c,
 		snapshotter:        rw.snapshotter,
 		diffID:             diffID,
 		contentStoreDigest: desc.Digest,
+		lease:              &lease,
 	}, nil
 }
 
 func (rw *rwlayer) Release() error {
-	snapshotter := rw.c.SnapshotService(rw.snapshotter)
-	err := snapshotter.Remove(context.TODO(), rw.key)
-	if err != nil && !cerrdefs.IsNotFound(err) {
-		return err
+	if rw.lease != nil {
+		lm := rw.c.LeasesService()
+		err := lm.Delete(context.TODO(), *rw.lease)
+		if err != nil {
+			log.G(context.TODO()).WithError(err).Warn("failed to delete lease when releasing RWLayer")
+		} else {
+			rw.lease = nil
+		}
 	}
-
 	if rw.root == "" { // nothing to release
 		return nil
 	}
```

Comment:
Definitely not a blocker, and only update if other changes are needed (slightly related to the above); 

Would we ever have a lease if there's no `root` ? If not then I'd consider putting this one as first condition of the function, as it's a very clear "return early, we're done".